### PR TITLE
GPU Setup for MPS

### PR DESCRIPTION
**Pull request by abhishekpratapa**
- The GPU Setup for MPS using the drawing classifier as an example.

### CODE REVIEW

**Comment by abhishekpratapa** [comment]
#1888

File: src/unity/python/turicreate/test/test_drawing_classifier.py [hunk 1]
```diff
@@ -323,3 +323,7 @@ def setUpClass(self):
         super(DrawingClassifierUsingQuickdraw245, self).setUpClass(
             warm_start="quickdraw_245_v0")
 
+class DrawingClassifierMPS(unittest.TestCase):
+    def test_cpp(self):
```
Comment (TobyRoseman) [inline src/unity/python/turicreate/test/test_drawing_classifier.py]: So this is a syntax error.

Also please add unit tests.

File: src/unity/toolkits/sound_classifier/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,9 @@
+project(unity_toolkits)
```
Comment (TobyRoseman) [inline src/unity/toolkits/sound_classifier/CMakeLists.txt]: Let's not worry about adding anything for the sound classifier to this review.

File: src/unity/toolkits/sound_classifier/sound_classifier.cpp [hunk 3]
```diff
@@ -0,0 +1 @@
+/* TODO: sound classifier class */
```
Comment (TobyRoseman) [inline src/unity/toolkits/sound_classifier/sound_classifier.cpp]: Also don't worry about this empty file.

File: src/unity/toolkits/drawing_classifier/drawing_classifier.mm [hunk 4]
```diff
@@ -0,0 +1,12 @@
+#include <unity/toolkits/drawing_classifier/drawing_classifier.hpp>
+#include <unity/toolkits/drawing_classifier/mps_drawing_classifier.hpp>
+
+namespace turi {
+namespace drawing_classifier {
+
+void init_model(std::map<std::string, std::vector<float>> weights) {
+
```
Comment (TobyRoseman) [inline src/unity/toolkits/drawing_classifier/drawing_classifier.mm]: I'm confused. Seems like there should be a lot more here before it's even code complete.

File: src/unity/toolkits/drawing_classifier/mps_drawing_classifier.mm [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+- (id _Nullable)initWithParameters:(NSString *_Nonnull)name
+                         inputNode:(MPSNNImageNode *_Nonnull)inputNode
+                            device:(id<MTLDevice> _Nonnull)dev
+                         cmd_queue:(id<MTLCommandQueue> _Nonnull)cmd_q
+                       initWeights:(std::map<std::string, std::vector<float>>)weights {
+  @autoreleasepool {
+    self = [super init];
+    return self;
+  }
+}
```
Comment (TobyRoseman) [inline src/unity/toolkits/drawing_classifier/mps_drawing_classifier.mm]: Also seems like this is missing a lot. I would expect to see the actual network definition here, but I'm not seeing that any where.

**Comment by abhishekpratapa** [comment]
> We talked about removing the classes which just wrap MPS layers, as these don't really add much. Are you still planning to remove these? (To be clear: I do agree we need the padding and data loading classes.)

Yes, I hadn't got to the refactor!

File: src/unity/toolkits/drawing_classifier/mps_drawing_classifier.mm (hunk 5 above)
Comment (abhishekpratapa) [reply src/unity/toolkits/drawing_classifier/mps_drawing_classifier.mm]: I forgot to push a commit.

Comment (abhishekpratapa) [reply src/unity/toolkits/drawing_classifier/mps_drawing_classifier.mm]: I pushed it's under the commit history but it's not showing up here. Hmm.

**Comment by nickjong** [comment]
After syncing with Abhishek offline, he'll take a pass at revising/refactoring:

1. (Minor) run `clang-format` to make some of the ObjC code stuff more readable
2. Make `mps_drawing_classifier`'s interface close to the `model_backend` interface (but in ObjC)
3. Revisit the question of wrapping each MPS node in its own layer class, with an eye towards making the actual graph definition as readable as possible

**Pull request closed without merging**